### ISSUE TITLE
Suggest `&string[6..11]` for `str::from_utf8(string.as_bytes()[6..11]).unwrap()`

### ISSUE THREAD

**Issue by Luro02** [rust-lang/rust-clippy#5487]
This

```rust
use std::str;

let string = "Hello World!";
assert_eq!(str::from_utf8(string.as_bytes()[6..11]).unwrap(), "World");
```

could be written as

```rust
let string = "Hello World!";
assert_eq!(&string[6..11], "World");
```

**Comment by l0calh05t** [comment]
```
assert_eq!(str::from_utf8(string.as_bytes()[6..11]).unwrap(), "World");
```

should be 

```
assert_eq!(str::from_utf8(&string.as_bytes()[6..11]).unwrap(), "World");
```

**Comment by illicitonion** [comment]
I believe these are only equivalent if string is only made of single-byte characters. Counter-example:
```rust
let string = "🐈ello World!";
assert_eq!(::std::str::from_utf8(&string.as_bytes()[6..11]).unwrap(), "World");
```

**Comment by Luro02** [comment]
Well of course your example fails. Neither `string[6..11]` nor `::std::str::from_utf8(&string.as_bytes()[6..11]).unwrap()` are equal to `"World"`.

The indexing works with bytes and not `char`s, which is why your example is not working as expected, because the emoji (`"🐈".len()`) is 4 bytes in length.

For example
```rust
let string = "🐈ello World!";

dbg!(&string[1..11]);
```

will fail with

```
thread 'main' panicked at 'byte index 1 is not a char boundary; it is inside '🐈' (bytes 0..4) of `🐈ello World!`', src/libcore/str/mod.rs:2219:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```


This is your example working correctly:
```rust
use std::str;

fn main() {
    let string = "🐈ello World!";

    assert_eq!(&string[6..11], "lo Wo");

    assert_eq!(
        str::from_utf8(&string.as_bytes()[6..11]).unwrap(),
        &string[6..11]
    );
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=use%20std%3A%3Astr%3B%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20string%20%3D%20%22%F0%9F%90%88ello%20World!%22%3B%0A%0A%20%20%20%20assert_eq!(%26string%5B6..11%5D%2C%20%22lo%20Wo%22)%3B%0A%0A%20%20%20%20assert_eq!(%0A%20%20%20%20%20%20%20%20str%3A%3Afrom_utf8(%26string.as_bytes()%5B6..11%5D).unwrap()%2C%0A%20%20%20%20%20%20%20%20%26string%5B6..11%5D%0A%20%20%20%20)%3B%0A%7D%0A)

The above code succeeds, which means that `str::from_utf8(&string.as_bytes()[6..11]).unwrap()` and `&string[6..11]` are equal, so the lint is also valid for utf8 strings.

**Comment by illicitonion** [comment]
> Well of course your example fails. Neither `string[6..11]` nor `::std::str::from_utf8(&string.as_bytes()[6..11]).unwrap()` are equal to `"World"`.
> 
> The indexing works with bytes and not `char`s, which is why your example is not working as expected, because the emoji (`"🐈".len()`) is 4 bytes in length.

Apologies, of course you're right - for some reason I was thinking slicing strings worked on `chars` and as you've shown I put together a really poor example to try to show that! Thanks for taking the time to make a better one which shows the truth :)
